File: macos/usdview/usd.swift
import Foundation

enum UsdPrimType {
    case mesh
    case xform
    case untyped

    init(rawString: String) {
        switch rawString {
        case "Mesh":
            self = .mesh
        case "Xform":
            self = .xform
        default:
            self = .untyped
        }
    }
}

struct UsdPrimInfo {
    let name: String
    let path: SdfPath
    let type: UsdPrimType
    let depth: Int

    init(path: SdfPath, type: UsdPrimType) {
        self.path = path
        name = path.getName()
        depth = path.getDepth()
        self.type = type
    }
}

enum UsdStageInterpolation {
    case held
    case linear
}

struct UsdStageInfo {
    let startTimeCode: Double
    let endTimeCode: Double
    let timeCodesPerSecond: Double
    let interpolationType: UsdStageInterpolation
    internal let ptr: UnsafeMutableRawPointer

    init(stagePath: String) {
        ptr = usdffi_openStage(stagePath)
        startTimeCode = usdffi_getStartTimeCode(ptr)
        endTimeCode = usdffi_getEndTimeCode(ptr)
        timeCodesPerSecond = usdffi_getTimeCodesPerSecond(ptr)

        let interpStr = String(cString: usdffi_getInterpolationType(ptr))
        if interpStr == "Held" {
            interpolationType = UsdStageInterpolation.held
        } else {
            interpolationType = UsdStageInterpolation.linear
        }
    }

    func getPrims() -> [UsdPrimInfo] {
        var numPrims: Int32 = 0
        var result: [UsdPrimInfo] = []
        var stream = usdffi_getPrimInfo(ptr, &numPrims)!

        let getString = { (ptr: inout UnsafeMutablePointer<UnsafePointer<Int8>?>) -> String in
            if let raw = ptr.pointee { ptr += 1; return String(cString: raw) }
            return ""
        }

        for _ in 0 ..< numPrims {
            let path = getString(&stream)
            let type = getString(&stream)

            result.append(UsdPrimInfo(path: SdfPath(rawString: path,
                                                    isPrimPath: true),
                                      type: UsdPrimType(rawString: type)))
        }

        return result
    }

    func reload() {
        usdffi_reloadStage(ptr)
    }

    func setDefaultPrim(path: SdfPath) {
        usdffi_setDefaultPrim(ptr, path.getString())
    }

    func getDefaultPrim() -> String {
        return String(cString: usdffi_getDefaultPrim(ptr))
    }
}

File: macos/usdview/usd.cpp
#include "usd.hpp"

#include "pxr/pxr.h"
#include "pxr/usd/sdf/declareHandles.h"
#include "pxr/usd/usd/prim.h"
#include "pxr/usd/usd/primRange.h"
#include "pxr/usd/usd/stage.h"

#include "boost/get_pointer.hpp"

auto allocStrBuffer = [](auto s) {
  return (const char **)malloc(s * sizeof(const char *));
};

auto reallocAndCopy = [](auto &oldStream, auto size) {
  const char **newStream = allocStrBuffer(2 * size);
  const char **oldStreamHandle = oldStream;
  for (size_t i = 0; i < size; ++i) {
    newStream[i] = oldStream[i];
  }

  oldStream = newStream;
  return oldStreamHandle;
};

// Exposes a stream of prim information in the form of
// [char*] path [char*] type
const char **usdffi_getPrimInfo(void *stagePtr, int *numPrims) {
  auto stage = (pxr::UsdStage *)stagePtr;
  const size_t numElementsPerStream = 2;
  size_t numElementsInStreamTotal = 32;
  const char **stream = allocStrBuffer(numElementsInStreamTotal);
  size_t currentStreamSize = 0;

  for (const auto &child : stage->Traverse()) {
    // We may have outgrown our original buffer estimate, if necessary
    // we will reallocate and copy our elements over, doubling the size.
    if (currentStreamSize < numElementsPerStream) {
      auto oldStream = reallocAndCopy(stream, numElementsInStreamTotal);
      numElementsInStreamTotal *= 2;
      free(oldStream);
    }

    stream[currentStreamSize++] = child.GetPath().GetText();
    stream[currentStreamSize++] = child.GetTypeName().GetText();

    *numPrims += 1;
  }
  return stream;
}

double usdffi_getStartTimeCode(void *stagePtr) {
  return ((pxr::UsdStage *)stagePtr)->GetStartTimeCode();
}

double usdffi_getEndTimeCode(void *stagePtr) {
  return ((pxr::UsdStage *)stagePtr)->GetEndTimeCode();
}

const char *usdffi_getInterpolationType(void *stagePtr) {
  auto interp = ((pxr::UsdStage *)stagePtr)->GetInterpolationType();
  return interp == 0 ? "Held" : "Linear";
}

double usdffi_getTimeCodesPerSecond(void *stagePtr) {
  return ((pxr::UsdStage *)stagePtr)->GetTimeCodesPerSecond();
}

void *usdffi_openStage(const char *sPath) {
  static auto stage = pxr::UsdStage::Open(sPath);
  return boost::get_pointer(stage);
}

void usdffi_reloadStage(void *stagePtr) {
  ((pxr::UsdStage *)stagePtr)->Reload();
}

void usdffi_setDefaultPrim(void *stagePtr, const char *primPath) {
  auto path = pxr::SdfPath(primPath);
  auto prim = ((pxr::UsdStage *)stagePtr)->GetPrimAtPath(path);
  ((pxr::UsdStage *)stagePtr)->SetDefaultPrim(prim);
}

const char *usdffi_getDefaultPrim(void *stagePtr) {
  return ((pxr::UsdStage *)stagePtr)->GetDefaultPrim().GetPath().GetText();
}
